File: bindings/imgui_bundle/demos_cpp/demos_immvision/demo_contours.cpp
// Adaptation of https://docs.opencv.org/4.x/df/d0d/tutorial_find_contours.html

#include "imgui.h"
#include "immapp/immapp.h"
#include "immvision/immvision.h"
#include "demo_utils/api_demos.h"

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>


struct BlurData
{
    cv::Mat src_gray;
    cv::Mat contours;
    int thresh = 100;

    BlurData(const std::string& imageFile)
    {
        auto& self = *this;
        cv::Mat img = cv::imread(imageFile);
        cv::resize(img, img, cv::Size(), 0.5, 0.5);
        cv::cvtColor(img, self.src_gray, cv::COLOR_BGR2GRAY);
        self.update_contours();
    }

    void update_contours()
    {
        auto& self = *this;
        cv::RNG rng = cv::RNG(12345);
        cv::Mat canny_output;
        Canny(self.src_gray, canny_output, self.thresh, self.thresh * 2 );
        std::vector<std::vector<cv::Point> > contours;
        std::vector<cv::Vec4i> hierarchy;
        findContours( canny_output, contours, hierarchy, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE );
        self.contours = cv::Mat::zeros(canny_output.size(), CV_8UC3 );
        for( size_t i = 0; i< contours.size(); i++ )
        {
            cv::Scalar color = cv::Scalar(rng.uniform(0, 256), rng.uniform(0,256), rng.uniform(0,256) );
            cv::drawContours(self.contours, contours, (int)i, color, 2, cv::LINE_8, hierarchy, 0 );
        }

    }
};


void gui_blur(BlurData& blur_data)
{
    const int min_thresh = 20, max_tresh = 500;
    cv::Size imageDisplaySize(400, 0);
    bool changed = ImGui::SliderInt("Canny thresh", &blur_data.thresh, min_thresh, max_tresh);
    if (changed)
        blur_data.update_contours();

    ImmVision::ImageDisplay("img", blur_data.src_gray, imageDisplaySize);
    ImGui::SameLine();
    ImmVision::ImageDisplay("contours", blur_data.contours, imageDisplaySize, changed);
};


void demo_contours()
{
    static BlurData blur_data(DemosAssetsFolder() + "/images/house.jpg");
    gui_blur(blur_data);
}


#ifndef IMGUI_BUNDLE_BUILD_DEMO_AS_LIBRARY
int main(int, char**)
{
    ImmApp::Run(demo_contours, "blur", true);
    //ImmApp::Run(demo_contours, "blur", false, false, {800, 600});
    return 0;
}
#endif

File: bindings/imgui_bundle/demos_python/demos_immvision/demo_contours.py
# Adaptation of https://docs.opencv.org/4.x/df/d0d/tutorial_find_contours.html

from imgui_bundle import imgui, immapp, immvision
from imgui_bundle.demos_python import demo_utils
import cv2  # type: ignore
import random as rng
import numpy as np
from numpy.typing import NDArray


class BlurData:
    src_gray: NDArray[np.uint8]
    contours: NDArray[np.uint8]
    thresh: int = 100

    def __init__(self, image_file: str):
        img = cv2.imread(image_file)
        img = cv2.resize(img, dsize=None, fx=0.5, fy=0.5)  # type: ignore
        self.src_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        self.update_contours()

    def update_contours(self):
        canny_output = cv2.Canny(self.src_gray, self.thresh, self.thresh * 2)
        contours, hierarchy = cv2.findContours(
            canny_output, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE
        )
        self.contours = np.zeros(
            (canny_output.shape[0], canny_output.shape[1], 3), dtype=np.uint8
        )
        for i in range(len(contours)):
            color = (rng.randint(0, 256), rng.randint(0, 256), rng.randint(0, 256))
            cv2.drawContours(
                self.contours, contours, i, color, 2, cv2.LINE_8, hierarchy, 0
            )


def gui_blur(blur_data: BlurData):
    min_thresh = 20
    max_thresh = 500
    image_display_size = (400, 0)
    changed, blur_data.thresh = imgui.slider_int(
        "Canny thresh", blur_data.thresh, min_thresh, max_thresh
    )
    if changed:
        blur_data.update_contours()
    immvision.image_display("img", blur_data.src_gray, image_display_size)
    imgui.same_line()
    immvision.image_display("contours", blur_data.contours, image_display_size, changed)


def main():
    blur_data = BlurData(demo_utils.demos_assets_folder() + "/images/house.jpg")

    def gui():
        gui_blur(blur_data)

    immapp.run(gui, "blur", True)


if __name__ == "__main__":
    main()
